src/v2/token.rs

```diff
@@ -0,0 +1,167 @@
+use std::sync::Arc;
+
+use cache_control::CacheControl;
+use josekit::{jwk::JwkSet, jws::RS256, jwt, jwt::JwtPayload};
+use reqwest::{header, Client, Response};
+use time::OffsetDateTime;
+use tokio::sync::RwLock;
+use url::Url;
+
+/// Zitadel client to verify a token's validity
+#[derive(Debug, Clone)]
+pub struct ZitadelJWTVerifier {
+	/// Zitadel domain
+	domain: Url,
+	/// Client for performing the requests
+	client: Client,
+	/// Key set from Zitadel
+	jwks: Arc<RwLock<JwkSet>>,
+	/// Time when the jwks is no longer valid
+	expires_at: OffsetDateTime,
+}
+
+impl ZitadelJWTVerifier {
+	/// Creates a new verifier to verify with a specific server
+	#[must_use]
+	pub fn new(url: Url) -> Self {
+		let jwks = JwkSet::new();
+		Self {
+			domain: url,
+			client: Client::new(),
+			jwks: Arc::new(RwLock::new(jwks)),
+			expires_at: OffsetDateTime::now_utc(),
+		}
+	}
+
+	/// Verifies if a token is valid and returns the token payload
+	/// The performed verifications are:
+	///     - Token signature
+	///     - Token not expired
+	///     - Token not used before 'not before'
+	///     - Token issuer is the expected server
+	pub async fn verify(&mut self, token: String) -> Result<JwtPayload, TokenValidationError> {
+		use TokenValidationError::*;
+
+		let header = jwt::decode_header(&token)?;
+		let kid = header
+			.claim("kid")
+			.ok_or(BadToken("No kid"))?
+			.as_str()
+			.ok_or(BadToken("kid is not a string"))?;
+
+		let mut jwk = self.jwks.read().await.get(kid).first().map(|&jwk| jwk.clone());
+		if self.expires_at < OffsetDateTime::now_utc() || jwk.is_none() {
+			let (new_jwks, expires_at) = self.get_jwks().await?;
+			let mut jwks = self.jwks.write().await;
+			*jwks = new_jwks;
+			jwk = jwks.get(kid).first().map(|&jwk| jwk.clone());
+			self.expires_at = expires_at;
```

Comment:
This is the only reason why `.verify` takes `&mut self`, should we maybe put this field  under the lock along with `jwks`? Othwerwise I'm afraid that a user of `ZitadelJWTVerifier` would have to wrap it in their own lock to use in async setting

Reply:
You are completely right, this would cause me problems later. Thanks for catching that. I made the changes but I couldn't think in a way to read the lock only once and don't have to drop the guard myself. Do you have any suggestions @sirewix ?

Reply:
Smth like this, but it doesn't add that much
```rust
		let jwks_cache = self.jwks_cache.read().await;
		let mut jwk = jwks_cache.jwks.get(kid).first().cloned().cloned();
		let expires_at = jwks_cache.expires_at;
		if expires_at < OffsetDateTime::now_utc() || jwk.is_none() {
			let (new_jwks, expires_at) = self.get_jwks().await?;
			let mut jwks_cache = self.jwks_cache.write().await;
			jwks_cache.jwks = new_jwks;
			jwks_cache.expires_at = expires_at;
			jwk = jwks_cache.jwks.get(kid).first().cloned().cloned();
			tracing::debug!("Updated JWKs");
		}
```

Reply:
Ah, yes, we need to limit guard's scope:
```rust
		let (mut jwk, expires_at) = {
			let jwks_cache = self.jwks_cache.read().await;
			( jwks_cache.jwks.get(kid).first().cloned().cloned(),
			  jwks_cache.expires_at.clone()
			)
		};
		if expires_at < OffsetDateTime::now_utc() || jwk.is_none() {
			let (new_jwks, expires_at) = self.get_jwks().await?;
			let mut jwks_cache = self.jwks_cache.write().await;
			jwks_cache.jwks = new_jwks;
			jwks_cache.expires_at = expires_at;
			jwk = jwks_cache.jwks.get(kid).first().cloned().cloned();
			tracing::debug!("Updated JWKs");
		}

```

Reply:
Also, if we make `get_jwks()` return `JwkSetCache`, few more lines can be saved:
```rust
		if expires_at < OffsetDateTime::now_utc() || jwk.is_none() {
			let mut jwks_cache = self.jwks_cache.write().await;
			self.get_jwks().await?.clone_into(&mut jwks_cache);
			jwk = jwks_cache.jwks.get(kid).first().cloned().cloned();
			tracing::debug!("Updated JWKs");
		}
```
I moved write lock acquisition one line up, as it might so happen that many different threads would start fetching jwks at the same and then sequentially writing them which is just useless. If one thread is doing token renewal, other threads should just wait till it's done. We want time window between the read and write lock acquisitions as small ass possible (or maybe no window at all, but that's a task for improvement maybe in the future)

Reply:
Thinking about it, scoping reading wouldn't achieve much anyway, maybe this is enough for now and we can improve later:
```rust
		let jwks_cache = self.jwks_cache.read().await;
		let mut jwk = jwks_cache.jwks.get(kid).first().cloned().cloned();
		let expires_at = jwks_cache.expires_at;
		if expires_at < OffsetDateTime::now_utc() || jwk.is_none() {
			let mut jwks_cache = self.jwks_cache.write().await;
			*jwks_cache = self.get_jwks().await?;
			jwk = jwks_cache.jwks.get(kid).first().cloned().cloned();
			tracing::debug!("Updated JWKs");
		}
		drop(jwks_cache);
```

Reply:
This lets option I think won't work, there we drop the guard after we try to lock for writing. I think this would cause problems. I liked the solution with the guard and a small scope. I always forget that I can return things from this small scopes.